api/src/recon.rs

```diff
@@ -0,0 +1,170 @@
+use anyhow::Result;
+use async_trait::async_trait;
+use ceramic_core::EventId;
+use recon::{AssociativeHash, Key, Sha256a};
+use serde::{Deserialize, Serialize};
+
+use crate::ResumeToken;
+
+#[async_trait]
+pub trait ReconInterest: Clone + Send + Sync {
+    type Key: Key;
+    type Hash: AssociativeHash + std::fmt::Debug + Serialize + for<'de> Deserialize<'de>;
+
+    async fn insert(&self, key: Self::Key) -> Result<()>;
+    async fn range(
+        &self,
+        start: Self::Key,
+        end: Self::Key,
+        offset: usize,
+        limit: usize,
+    ) -> Result<Vec<Self::Key>>;
+}
+
+#[async_trait]
+pub trait ReconModel: Clone + Send + Sync {
+    type Key: Key;
+    type Hash: AssociativeHash + std::fmt::Debug + Serialize + for<'de> Deserialize<'de>;
+
+    async fn insert(&self, key: Self::Key, value: Option<Vec<u8>>) -> Result<()>;
+    async fn range_with_values(
+        &self,
+        start: Self::Key,
+        end: Self::Key,
+        offset: usize,
+        limit: usize,
+    ) -> Result<Vec<(Self::Key, Vec<u8>)>>;
+
+    async fn value_for_key(&self, key: Self::Key) -> Result<Option<Vec<u8>>>;
+
+    async fn keys_since_highwater_mark(
+        &self,
+        highwater: HighwaterMark,
+        limit: usize,
+    ) -> Result<(ResumeToken, Vec<Self::Key>)>;
+}
+
+#[async_trait::async_trait]
+pub trait AccessModelStore: Clone + Send + Sync {
+    type Key: recon::Key;
+
+    async fn keys_since_highwater_mark(
+        &self,
+        highwater: HighwaterMark,
+        limit: i64,
+    ) -> anyhow::Result<(String, Vec<Self::Key>)>;
+}
+
+#[derive(Clone, Debug)]
+pub struct HighwaterMark(i64);
```

Comment:
Why are there two types? HighwaterMark and ResumeToken? I see we can convert between them but what is the semantic difference?

Reply:
Good question.. I think highwater mark can be deleted. I originally wanted to make the resume token/cursor something that might apply to multiple routes and be the public serializable type, and then use the highwater mark internally to wrap the storage system's type.. but I didn't implement it well and having both types in the one interface is a mistake.